### PR TITLE
Playbook for CIS 1.16

### PR DESCRIPTION
**Pull request by maheel**

### CODE REVIEW

File: source/playbooks/CIS/lib/cis-stack.ts [hunk 1]
```diff
@@ -106,6 +106,29 @@ export class CisStack extends cdk.Stack {
...
+    //---------------------------------------------------------------------
+    // CIS116
+    //
+    let cis116findings = [
+      "1.16 Ensure IAM policies are attached only to groups or roles",
+    ];
+
+    const cis116playbook: PlaybookConstruct = new PlaybookConstruct(this, 'cis116playbook', {
+      name: 'CIS116',
+      description: 'Remediates CIS 1.16 by adding the users to a group with relevant polices, removing policies attached to them.',
```
Comment (gergnz) [inline source/playbooks/CIS/lib/cis-stack.ts]: maybe:
`Remediates CIS 1.16 by adding users to a new group with polices obtained from the users.`

File: source/playbooks/CIS/lambda/cis116.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        details = finding.details['Resources'][0]['Details']
+        print(details)
+
+        aws_iam_user = details['AwsIamUser']
+        managed_policies = aws_iam_user['AttachedManagedPolicies'] if 'AttachedManagedPolicies' in aws_iam_user else []
+        username = aws_iam_user['UserName']
+        user_groups = aws_iam_user['GroupList'] if 'GroupList' in aws_iam_user else []
+        inline_policies = aws_iam_user['UserPolicyList'] if 'UserPolicyList' in aws_iam_user else []
+
+        group_name = f'{username}-sharr-cis116'
```
Comment (gergnz) [inline source/playbooks/CIS/lambda/cis116.py]: Good idea to use the extension `-sharr-cis116`, but it kind of leaves artefacts around.

Suggestion: check if a group exists with just the username, if not, then just use username, else add a random 8 character string, similar to how CloudFormation does with resource names, double check that doesn't exist, then use that.

File: source/playbooks/CIS/lib/cis-stack.ts (hunk 1 above)
Comment (maheel) [reply source/playbooks/CIS/lib/cis-stack.ts]: Ok

File: source/playbooks/CIS/lambda/cis116.py (hunk 2 above)
Comment (maheel) [reply source/playbooks/CIS/lambda/cis116.py]: Ok

File: source/playbooks/CIS/lib/cis-stack.ts (hunk 1 above)
Comment (maheel) [reply source/playbooks/CIS/lib/cis-stack.ts]: Fixed

File: source/playbooks/CIS/lambda/cis116.py (hunk 2 above)
Comment (maheel) [reply source/playbooks/CIS/lambda/cis116.py]: Fixed

File: source/playbooks/CIS/lambda/cis116.py [hunk 3]
```diff
@@ -190,19 +193,46 @@ def failed():
...
         notify(finding, message, LOGGER, cwlogs=APPLOGGER, sns=AWS)
         return
 
     except Exception as e:
         LOGGER.error(e)
         failed()
         return
+
+
+def does_group_exist(iam, group_name):
```
Comment (gergnz) [inline source/playbooks/CIS/lambda/cis116.py:226]: This function should go above where it is called from or you will get a `NameError` not found when you try to call it on line 140.

File: source/playbooks/CIS/lambda/cis120.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        notify(finding, message, LOGGER, cwlogs=APPLOGGER, sns=AWS)
+        return
+
+    except Exception as e:
+        LOGGER.error(e)
+        failed()
+        return
+
+
+def does_role_exist(iam, role_name):
```
Comment (gergnz) [inline source/playbooks/CIS/lambda/cis120.py:193]: This function needs to go above where it is called or you will get a `NameError` not found.

File: source/playbooks/CIS/lambda/cis120.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+        """
+    role_exists = False
+
+    try:
+        response = iam.get_role(
+            RoleName=role_name
+        )
+
+        if 'Role' in response:
+            print('Yes exists')
```
Comment (gergnz) [inline source/playbooks/CIS/lambda/cis120.py]: Should probably change this to `LOGGER.debug()`

Comment (maheel) [reply source/playbooks/CIS/lambda/cis120.py]: This was for debgging. Will remove, forgot to remove.

File: source/playbooks/CIS/lambda/cis120.py (hunk 4 above)
Comment (maheel) [reply source/playbooks/CIS/lambda/cis120.py:193]: I doubt if this is the case, as in Python you can define in any order. did you confuse with JS? I tested both locally and playbook actions.

File: source/playbooks/CIS/lambda/cis116.py (hunk 3 above)
Comment (maheel) [reply source/playbooks/CIS/lambda/cis116.py:226]: I doubt if this is the case, as in Python you can define in any order. did you confuse with JS? I tested both locally and playbook actions.

File: source/playbooks/CIS/lambda/cis120.py (hunk 4 above)
Comment (gergnz) [reply source/playbooks/CIS/lambda/cis120.py:193]: I swear it's python, but no probs. (maybe it used to be the case...)

**Pull request merged**